### PR TITLE
rg-62: user email confirmation

### PR DESCRIPTION
**Pull request by MariaPetrukhnova**
Add main logic:
1. Submit sign up data
2. Get it on back and check, send verification email
3. Create email confirmation via emailService
4. Confirm email with click on link in email
5. Accept token on confirmed-email route
6. Add to database new user with sign up data and emailConfirmed: true for further restriction in login

I know, it can have a lot of change requests because there are a lot to do further: rediretion, login proceeding. But I need here some comments from you, because it was rather challenging task for me with entry code in our project

### CODE REVIEW

**Comment by Vitaliy-Kalenichenko** [comment]
Conflicts :(

**Comment by Vitaliy-Kalenichenko** [comment]
Conflicts :(

File: backend/src/bundles/auth/auth.controller.ts [hunk 1]
```diff
@@ -32,6 +34,11 @@ import { type ILogger } from '~/common/logger/logger.js';
 
 import { type AuthService } from './auth.service.js';
 
+interface JwtPayload {
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/auth/auth.controller.ts:37]: Consider changing to a type

image

File: backend/src/bundles/auth/auth.controller.ts [hunk 2]
```diff
@@ -317,6 +323,34 @@ class AuthController extends Controller {
...
+            const decodedToken = jwt.verify(
+                token,
+                config.ENV.JWT.ACCESS_TOKEN_SECRET,
+            ) as JwtPayload;
+            // Call confirmUserEmail function with the decoded token
+            await userService.confirmUserEmail(decodedToken);
+
+            return {
+                status: HttpCode.OK,
+                payload: { message: 'Email confirmed successfully' },
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/auth/auth.controller.ts:341]: Consider moving to separate email messages enum

File: backend/src/bundles/auth/auth.service.ts [hunk 3]
```diff
@@ -46,11 +56,14 @@ class AuthService implements TAuthService {
...
         );
 
-        const { id } = await this.userService.create({
+        const { id, email } = await this.userService.create({
             ...userRequestDto,
             passwordSalt,
             passwordHash,
         });
+        const token = generateToken({ id });
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/auth/auth.service.ts:65]: Consider refactoring without disabling rules

File: backend/src/bundles/auth/auth.service.ts [hunk 4]
```diff
@@ -61,6 +74,28 @@ class AuthService implements TAuthService {
...
+                title: 'ResumeGemm',
+                dashboardLink: verificationLink,
+                logoLink: this.config.ENV.EMAIL.SMTP_LOGO,
+            },
+        });
+
+        await mailService.sendMail({
+            to: email,
+            subject: 'You have successfully registred on ResumeGemm',
+            text: 'You have successfully registred on ResumeGemm',
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/auth/auth.service.ts:94]: Consider moving to enum

File: backend/src/bundles/users/user.service.ts [hunk 5]
```diff
@@ -12,6 +12,11 @@ import {
     type UserWithProfileRelation,
 } from './types/types.js';
 
+interface JwtPayload {
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/users/user.service.ts:15]: Repeats in auth.controller, consider moving to a separate file

File: frontend/src/bundles/auth/components/comfirmed-email-page/confirmed-email-page.tsx [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import { useEffect, useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+
+import { Logo } from '../components';
+import styles from './styles.module.scss';
+
+const EmailConfirmedPage = (): JSX.Element => {
+    const [counter, setCounter] = useState(15);
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/comfirmed-email-page/confirmed-email-page.tsx:8]: Move 15 to a constant

File: frontend/src/bundles/auth/components/comfirmed-email-page/confirmed-email-page.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    const navigate = useNavigate();
+
+    useEffect(() => {
+        const timer = setInterval(() => {
+            setCounter((previousCounter) => previousCounter - 1);
+        }, 1000);
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/comfirmed-email-page/confirmed-email-page.tsx:14]: Move to a constant

File: frontend/src/bundles/auth/components/sign-up-form/components/styles.module.scss [hunk 8]
```diff
@@ -0,0 +1,57 @@
+.input {
+    height: 37px;
+    padding: 12px;
+    color: #434c66;
+    font-size: 14px;
+    line-height: 20px;
+    background: #ebeff1;
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/sign-up-form/components/styles.module.scss:7]: consider replacing with css vars

File: frontend/src/bundles/auth/components/sign-up-form/constants/constants.ts [hunk 9]
```diff
@@ -1,10 +1,19 @@
...
-const DEFAULT_SIGN_UP_PAYLOAD: UserSignUpRequestDto = {
+import { type UserSignUpRequestDtoFrontend } from '../validation/sign-up-validation';
+
+const DEFAULT_SIGN_UP_PAYLOAD: UserSignUpRequestDtoFrontend = {
     firstName: '',
     lastName: '',
     email: '',
     password: '',
+    confirm_password: '',
+    // emailConfirmed: false
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/sign-up-form/constants/constants.ts:11]: image

File: frontend/src/bundles/auth/components/sign-up-form/sign-up-form.tsx [hunk 10]
```diff
@@ -40,8 +46,12 @@ const SignUpForm: React.FC<Properties> = ({ onSubmit }) => {
     const handleFormSubmit = useCallback(
         (event_: React.BaseSyntheticEvent): void => {
             void handleSubmit(onSubmit)(event_);
+            showToast(
+                'Thank you for sign up! Check please your email to confirm registration',
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/sign-up-form/sign-up-form.tsx:50]: consider moving to enum

File: frontend/src/bundles/auth/components/sign-up-form/validation/sign-up-validation.ts [hunk 11]
```diff
@@ -9,7 +9,8 @@ type UserSignUpRequestDtoFrontend = {
     lastName: string;
     email: string;
     password: string;
-    confirm_password?: string;
+    confirm_password: string;
+    // emailConfirmed: boolean;
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/sign-up-form/validation/sign-up-validation.ts:13]: image

File: frontend/src/bundles/auth/components/sign-up-form/validation/sign-up-validation.ts [hunk 12]
```diff
@@ -57,6 +58,7 @@ const userSignUpValidationFrontend = joi.object<
             'any.only': UserValidationMessage.CONFIRM_PASSWORD_MATCH,
             'string.empty': UserValidationMessage.CONFIRM_PASSWORD_REQUIRED,
         }),
+    // emailConfirmed: joi.boolean()
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/auth/components/sign-up-form/validation/sign-up-validation.ts:61]: image

File: frontend/src/bundles/payment/pages/subscription-payment.tsx [hunk 13]
```diff
@@ -37,6 +36,7 @@ const SubscriptionPaymentPage: React.FC = () => {
 
     const [modalIsHidden, setModalIsHidden] = useState(true);
 
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/payment/pages/subscription-payment.tsx:39]: image

File: frontend/src/bundles/users/users-api.ts [hunk 14]
```diff
@@ -29,6 +32,19 @@ class UserApi extends HttpApi {
...
+    public async loadUser(): Promise<UserSignUpResponseDto> {
+        const response = await this.load(
+            this.getFullEndpoint(UsersApiPath.GET_USER, {}),
+            {
+                method: 'GET',
+                contentType: ContentType.JSON,
+                hasAuth: true,
+            },
+        );
+        setTimeout(() => false, 5000);
```
Comment (MaxPoplavskiy) [inline frontend/src/bundles/users/users-api.ts:45]: consider moving to a constant

File: frontend/src/framework/storage/enums/storage-key.enum.ts [hunk 15]
```diff
@@ -1,5 +1,6 @@
 enum StorageKey {
     ACCESS_TOKEN = 'access_token',
+    NAME_EXIST = 'nameExist',
```
Comment (MaxPoplavskiy) [inline frontend/src/framework/storage/enums/storage-key.enum.ts:3]: consider using the same case

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/users/users-api.ts (hunk 14 above)
Comment (MaxPoplavskiy) [reply frontend/src/bundles/users/users-api.ts:45]: Isn't this setTimeout redundant?

File: backend/src/bundles/auth/enums/message.enum.ts [hunk 16]
```diff
@@ -0,0 +1,6 @@
+const userMessages = {
+    EMAIL_CONFIRMED: 'Email confirmed successfully',
+    SUCCESSFULLY_REGISTRED: 'You have successfully registred on ResumeGemm',
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/auth/enums/message.enum.ts:3]: Word misspelled(REGISTRED)

File: backend/src/bundles/auth/enums/message.enum.ts [hunk 17]
```diff
@@ -0,0 +1,6 @@
+const userMessages = {
```
Comment (MaxPoplavskiy) [inline backend/src/bundles/auth/enums/message.enum.ts:1]: rewrite as ts enum

File: backend/src/common/database/enums/database-column-name.enum.ts [hunk 18]
```diff
@@ -62,7 +62,6 @@ enum DatabaseColumnName {
     LANGUAGE_LEVEL = 'language_level',
     ACTIVITY = 'activity',
     OWNER_ID = 'owner_id',
-    DESCRIPTION = 'description',
```
Comment (MaxPoplavskiy) [inline backend/src/common/database/enums/database-column-name.enum.ts:65]: Wouldn't it crash our dev branch?

File: frontend/src/framework/api/http-api.ts [hunk 19]
```diff
@@ -117,6 +117,7 @@ class HTTPApi implements IHttpApi {
 
         const isCustomException = Boolean(parsedException.errorType);
 
+        // eslint-disable-next-line @typescript-eslint/no-throw-literal
```
Comment (MaxPoplavskiy) [inline frontend/src/framework/api/http-api.ts:120]: image

File: shared/src/bundles/users/types/user-sign-up-request-dto.type.ts [hunk 20]
```diff
@@ -3,6 +3,8 @@ type UserSignUpRequestDto = {
     lastName: string;
     email: string;
     password: string;
+    confirm_password?: string;
```
Comment (MaxPoplavskiy) [inline shared/src/bundles/users/types/user-sign-up-request-dto.type.ts:6]: consider using camelCase

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]

File: backend/src/bundles/auth/auth.controller.ts [hunk 21]
```diff
@@ -209,11 +221,11 @@ class AuthController extends Controller {
         }>,
     ): Promise<ApiHandlerResponse<UserSignUpResponseDto>> {
         try {
-            const payload = await this.authService.signUp(options.body);
+            const token = await this.authService.signUp(options.body);
```
Comment (Icerock10) [inline backend/src/bundles/auth/auth.controller.ts:224]: are you sure this line returns only token?

File: backend/src/bundles/auth/auth.controller.ts [hunk 22]
```diff
@@ -318,6 +330,46 @@ class AuthController extends Controller {
...
+        options: ApiHandlerOptions<{ body: UserConfirmEmailRequestDto }>,
+    ): Promise<ApiHandlerResponse<UserConfirmEmailResponse>> {
+        try {
+            const { emailConfirmToken } = options.body;
+
+            if (!emailConfirmToken) {
+                return {
+                    status: HttpCode.BAD_REQUEST,
+                    payload: {
+                        message: 'Email confirm token not found',
```
Comment (Icerock10) [inline backend/src/bundles/auth/auth.controller.ts:344]: Write an exception message for this

File: backend/src/bundles/auth/helpers/get-template.ts [hunk 23]
```diff
@@ -0,0 +1,23 @@
+import fs from 'node:fs';
+import path, { dirname } from 'node:path';
+import { fileURLToPath } from 'node:url';
+
+import * as Pug from 'pug';
+
+type Properties = {
+    name: string;
+    context: object;
```
Comment (Icerock10) [inline backend/src/bundles/auth/helpers/get-template.ts:9]: `context` contains certain fields

File: backend/src/bundles/users/user.repository.ts [hunk 24]
```diff
@@ -43,6 +43,16 @@ class UserRepository
...
 
+    public async updateById(
+        id: string,
+        data: object,
+    ): Promise<UserEntity | null> {
+        await this.model.query().update(data).where('id', id).execute();
+
+        const user = await this.model.query().findOne({ id });
+        return user ? UserEntity.initialize(user) : null;
+    }
```
Comment (Icerock10) [inline backend/src/bundles/users/user.repository.ts:54]: Maria wrote this method, however, it has 0 usage across the app, maybe remove it?

File: backend/src/bundles/users/user.service.ts [hunk 25]
```diff
@@ -114,6 +116,10 @@ class UserService
         ) as Promise<UserWithProfileRelation>;
     }
 
+    public async confirmEmail(id: string): Promise<void> {
+        return await this.userRepository.confirmEmail(id);
```
Comment (Icerock10) [inline backend/src/bundles/users/user.service.ts:120]:
```suggestion
        return this.userRepository.confirmEmail(id);
```

File: frontend/src/bundles/auth/components/confirmed-email-page/confirmed-email-page.tsx [hunk 26]
```diff
@@ -0,0 +1,100 @@
+import { useCallback, useEffect, useState } from 'react';
```
Comment (Icerock10) [inline frontend/src/bundles/auth/components/confirmed-email-page/confirmed-email-page.tsx:1]:
```suggestion
import { useAppDispatch, useCallback, useEffect, useState } from '~/bundles/common/hooks/hooks';
```

**Review by Icerock10 — CHANGES_REQUESTED** [review]

**Comment by Icerock10** [comment]
As per the view, I would suggest to create a "Waiting page" like resume.io has, it's not a good ux to keep the user on the sign-up page, or we can let him register, and then ask to confirm email let's say in profile

image

**Review by vlad-blip — APPROVED** [review]

**Comment by vlad-blip** [comment]
[Add email confirmation](https://trello.com/c/UWw1OVVa/62-add-email-confirmation)

**Review by Icerock10 — APPROVED** [review]

File: backend/src/common/database/enums/database-column-name.enum.ts (hunk 18 above)
Comment (MaxPoplavskiy) [reply backend/src/common/database/enums/database-column-name.enum.ts:73]: Wouldn't it crash our dev branch?

**Comment by MaxPoplavskiy** [comment]
Wait for rg-180 to be merged

**Review by MaxPoplavskiy — APPROVED** [review]

File: frontend/src/bundles/auth/components/sign-up-form/components/styles.module.scss (hunk 8 above)
Comment (MaxPoplavskiy) [reply frontend/src/bundles/auth/components/sign-up-form/components/styles.module.scss:7]: consider replacing with css vars

**Review by MaxPoplavskiy — CHANGES_REQUESTED** [review]
Resolve old request changes

**Comment by ValentinaTenus** [comment]
> Resolve old request changes

they are resolved

**Review by MaxPoplavskiy — APPROVED** [review]

**Pull request merged**